operator/controllers/operation_controller.go

```diff
@@ -20,34 +20,338 @@ import (
 	ctrl "sigs.k8s.io/controller-runtime"
 	"sigs.k8s.io/controller-runtime/pkg/client"
-
-	operatorv1alpha1 "k8s.io/kubeadm/operator/api/v1alpha1"
 )
 
 // OperationReconciler reconciles a Operation object
 type OperationReconciler struct {
 	client.Client
-	Log logr.Logger
+	ManagerContainerName string
+	ManagerNamespace     string
+	AgentImage           string
+	MetricsRBAC          bool
+	Log                  logr.Logger
+	recorder             record.EventRecorder
 }
 
+// +kubebuilder:rbac:groups=core,resources=events,verbs=get;list;watch;create;patch
+// +kubebuilder:rbac:groups=core,resources=pods,verbs=get;list;watch;
+// +kubebuilder:rbac:groups=apps,resources=daemonsets,verbs=get;list;watch;create;update;patch;delete
 // +kubebuilder:rbac:groups=operator.kubeadm.x-k8s.io,resources=operations,verbs=get;list;watch;create;update;patch;delete
 // +kubebuilder:rbac:groups=operator.kubeadm.x-k8s.io,resources=operations/status,verbs=get;update;patch
 
+// SetupWithManager configures the controller for calling the reconciler
+func (r *OperationReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	err := ctrl.NewControllerManagedBy(mgr).
+		For(&operatorv1.Operation{}).
+		Owns(&operatorv1.RuntimeTaskGroup{}).
+		Complete(r)
+
+	//TODO: watch DS for operation Daemonsets
+
+	r.recorder = mgr.GetEventRecorderFor("operation-controller")
+	return err
+}
+
 // Reconcile an operation
-func (r *OperationReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
-	_ = context.Background()
-	_ = r.Log.WithValues("operation", req.NamespacedName)
+func (r *OperationReconciler) Reconcile(req ctrl.Request) (_ ctrl.Result, rerr error) {
+	ctx := context.Background()
+	log := r.Log.WithValues("operation", req.NamespacedName)
+
+	// Fetch the Operation instance
+	operation := &operatorv1.Operation{}
+	if err := r.Client.Get(ctx, req.NamespacedName, operation); err != nil {
+		if apierrors.IsNotFound(err) {
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, err
+	}
 
-	// your logic here
+	// Ignore the Operation if it is already completed or failed
+	if operation.Status.CompletionTime != nil {
+		// Reconcile the daemon set that deploys controller agents on nodes, so we are sure it is deleted after completion
+		err := r.reconcileDamonSet(operation, log)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+		return ctrl.Result{}, nil
+	}
+
+	// Initialize the patch helper
+
+	patchHelper, err := patch.NewHelper(operation, r)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	// Always attempt to Patch the Operation object and status after each reconciliation.
+	defer func() {
+		if err := patchHelper.Patch(ctx, operation); err != nil {
+			log.Error(err, "failed to patch Operation")
+			if rerr == nil {
+				rerr = err
+			}
+		}
+	}()
+
+	// Reconcile the Operation
+	if err := r.reconcileOperation(operation, log); err != nil {
+		if requeueErr, ok := errors.Cause(err).(capierrors.HasRequeueAfterError); ok {
```

Comment:
This is including `capi` types, do we want to do this?

Reply:
ATM we are using also other utils from CAPI; I would defer forking/duplicating this code to a later stage.